numpy/core/src/private/mem_overlap.c

```diff
@@ -0,0 +1,814 @@
+    stride[j] for some ordering.
+
+    In this case, for each x[i] at most one candidate exists, given that the
+    algorithm runs with strides sorted from largest to smallest. The problem can
+    be written as::
+
+       sum a_j x_j ?= b = sum a_j z_j
+
+       a_j = n_{j+1} * n_{j+2} * ... * n_d,  a_d = 1
+       0 <= x_j <= u_j <= 2*n_j - 2
+       0 <= z_j <= n_j - 1
+
+    b is the offset of the last element of the second array from the start of
+    the first.  z_j are uniquely determined because of the gcd property. For
+    each x_j, the bounds at first sight allow x_j=z_j and x_j=z_j+n_j. However,
+    u_j <= n_j - 1 + z_j, so that at most one candidate is left.
+
+  - Two arrays with stride-incommensurate starting points. For example,
+    a=x[:,::2], b=x[:,1::2].
+
+    The base address difference is incommensurate with all strides, so that
+    there are no solution candidates to consider. For itemsize != 1, similar
+    result is obtained for x_{n-1}.
+
+  The above cases cover arrays produced by typical slicing of well-behaved
+  parent arrays. More generally, more difficult cases can result::
+
+      x = np.arange(4*20).reshape(4, 20).astype(np.int8)
+      a = x[:,::7]
+      b = x[:,3::3]
+
+      <=>
+
+      20*x1 + 7*x2 + 3*x3 = 78    (= 3 + 3*20 + 5*3)
+      0 <= x1 <= 6, 0 <= x2 <= 2, 0 <= x3 <= 5
+
+  Non-overlapping in this case relies on x.shape[1] <= lcm(7, 3) = 21.  However,
+  elimination of x1 does not restrict candidate values for x3, so the algorithm
+  ends up considering all values x3=0...5 separately.
+
+  The upper bound for work done is prod(shape_a)*prod(shape_b), which scales
+  faster than than work done by binary ufuncs, after broadcasting,
+  prod(shape_a). The bound may be loose, but it is possible to construct hard
+  instances where ufunc is faster (adapted from [2,3])::
+
+      from numpy.lib.stride_tricks import as_strided
+      # Construct non-overlapping x1 and x2
+      x = np.zeros([192163377], dtype=np.int8)
+      x1 = as_strided(x, strides=(36674, 61119, 85569), shape=(1049, 1049, 1049))
+      x2 = as_strided(x[64023025:], strides=(12223, 12224, 1), shape=(1049, 1049, 1))
+
+  To avoid such worst cases, the amount of work done needs to be capped. If the
+  overlap problem is related to ufuncs, one suitable cap choice is to scale
+  max_work with the number of elements of the array. (Ref. [3] describes a more
+  efficient algorithm for solving problems similar to the above --- however,
+  also it must scale exponentially.)
+
+
+  *Integer overflows*
+
+  The algorithm is written in fixed-width integers, and can terminate with
+  failure if integer overflow is detected (the implementation catches all
+  cases). Potential failure modes:
+
+  - Array extent sum(stride*(shape-1)) is too large (for int64).
+
+  - Minimal solutions to a_i x_i + a_j x_j == b are too large,
+    in some of the intermediate equations.
```

Comment:
The magnitudes are bounded by |x1|<a2/gcd(a1,a2), |x2|<a1/gcd(a1,a2).
Overflows can then occur if both strides and end-start offset of the two arrays are of order 2**32. 

Example:
```
from numpy.core.multiarray_tests import solve_diophantine

p = 2**32 - 1  # Mersenne prime
a1 = p
a2 = p + p - 2

strides_a = (a1, 1)
shape_a = (3, 3*p)
offset_a = 1

strides_b = (a2, 1)
shape_b = (3, 2)
offset_b = p

rhs = offset_a - offset_b + sum(s*(n-1) for s,n in zip(strides_a, shape_a))

solve_diophantine(strides_a + strides_b,
                  tuple(n-1 for n in shape_a) + tuple(n-1 for n in shape_b),
                  rhs,
                  simplify=1)
```
Using 128-bit integers in the computation of t_l,t_u might avoid all integer overflows.